pkg/secrets/vault.go

```diff
@@ -2,23 +2,242 @@ package secrets
 
 import (
 	"fmt"
+	"github.com/hashicorp/vault/api"
+	"io/ioutil"
+	"os"
+	"strings"
 )
 
+type VaultConfig struct {
+	Enabled    bool   `json:"enabled" yaml:"enabled"`
+	Url        string `json:"url" yaml:"url"`
+	AuthMethod string `json:"authMethod" yaml:"authMethod"`
+	Role       string `json:"role" yaml:"role"`
+	Path       string `json:"path" yaml:"path"`
+	Token      string
+}
+
 type VaultSecret struct {
 	engine        string
-	namespace     string
+	path          string
 	key           string
-	base64Encoded bool
+	base64Encoded string
 }
 
 type VaultDecrypter struct {
-	secretConfig string
+	encryptedSecret string
+	vaultConfig     VaultConfig
+	isKV2           map[string]struct{}
+}
+
+func NewVaultDecrypter(encryptedSecret string) *VaultDecrypter {
+	return &VaultDecrypter{encryptedSecret, VaultConfig{}, map[string]struct{}{}}
 }
 
-func NewVaultDecrypter(secretConfig string) *VaultDecrypter {
-	return &VaultDecrypter{secretConfig}
+func getVaultDecrypter(encryptedSecret string) Decrypter {
+	decrypter := NewVaultDecrypter(encryptedSecret)
+	decrypter.vaultConfig = Registry.VaultConfig
+	return decrypter
 }
 
 func (v *VaultDecrypter) Decrypt() (string, error) {
-	return "", fmt.Errorf("vault secrets unsupported")
+	if err := v.ValidateVaultConfig(); err != nil {
```

Comment:
We shouldn't validate while decrypting (so for every secret) but when the config is set.

Reply:
Yes, good call. Changing